### PR TITLE
xdsclient: e2e style tests for ads stream restart (5/N)

### PR DESCRIPTION
**Pull request by easwars**
#a71-xds-fallback
#xdsclient-refactor

Addresses https://github.com/grpc/grpc-go/issues/6902

RELEASE NOTES: none

### CODE REVIEW

File: xds/internal/xdsclient/tests/ads_stream_restart_test.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+	// Create an xDS client with the above bootstrap configuration.
+	client, close, err := xdsclient.NewForTesting(xdsclient.OptionsForTesting{
+		Name:     t.Name(),
+		Contents: bootstrapContents,
+	})
+	if err != nil {
+		t.Fatalf("Failed to create xDS client: %v", err)
+	}
+	defer close()
```
Comment (zasweq) [inline xds/internal/xdsclient/tests/ads_stream_restart_test.go:120]: Optional: amongst PR's, these operations seem to be similar or the same across all tests. Should this be a helper instead?

File: xds/internal/xdsclient/tests/ads_stream_restart_test.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+	}
+
+	// Wait for a short duration and verify that no CDS request is sent, since
+	// there are no resources being watched.
+	sCtx, sCancel = context.WithTimeout(ctx, defaultTestShortTimeout)
+	defer sCancel()
+	select {
+	case <-sCtx.Done():
+	case names := <-cdsResourcesCh:
+		t.Fatalf("LDS request sent for resource names %v, when expecting no request", names)
```
Comment (zasweq) [inline xds/internal/xdsclient/tests/ads_stream_restart_test.go:233]: CDS

File: xds/internal/xdsclient/tests/ads_stream_restart_test.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+		select {
+		case <-ctx.Done():
+		case gotNames := <-namesCh:
+			if cmp.Equal(gotNames, wantNames, cmpopts.EquateEmpty(), cmpopts.SortSlices(func(s1, s2 string) bool { return s1 < s2 })) {
+				return nil
+			}
+			t.Logf("Received resource names %v, want %v", gotNames, wantNames)
+		}
+	}
+	return fmt.Errorf("timeout waiting for resource to be requested from the management server")
```
Comment (zasweq) [inline xds/internal/xdsclient/tests/ads_stream_restart_test.go:253]: Optional: should this also show the last resource names requested if any?

**Review by zasweq — APPROVED** [review]
Minor comments but overall LGTM.

File: xds/internal/xdsclient/tests/ads_stream_restart_test.go (hunk 1 above)
Comment (easwars) [reply xds/internal/xdsclient/tests/ads_stream_restart_test.go:120]: Yes, I was looking for things that could be replaced with helpers and creating the xDS client seemed like the only possible thing. But even there, some tests have slightly different way of creating the client like specifying an optional backoff, or specifying idle channel timeout etc. So, having a helper might not be so useful.

I'll probably revisit this once all tests are in and I can make a decision then.

File: xds/internal/xdsclient/tests/ads_stream_restart_test.go (hunk 2 above)
Comment (easwars) [reply xds/internal/xdsclient/tests/ads_stream_restart_test.go:233]: Done.

File: xds/internal/xdsclient/tests/ads_stream_restart_test.go (hunk 3 above)
Comment (easwars) [reply xds/internal/xdsclient/tests/ads_stream_restart_test.go:253]: Done.

File: xds/internal/xdsclient/tests/ads_stream_restart_test.go (hunk 1 above)
Comment (zasweq) [reply xds/internal/xdsclient/tests/ads_stream_restart_test.go:120]: Alright sounds good.

**Pull request merged**